### PR TITLE
Add sign function

### PR DESCRIPTION
**Pull request by hvy**
Added the sign function with gradient 0 everywhere, mostly because it could be used when reimplementing Functions as FunctionNodes in the backward pass.

### CODE REVIEW

File: chainer/functions/math/sign.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        type_check.expect(in_types.size() == 1)
+
+    def forward_cpu(self, x):
+        return utils.force_array(numpy.sign(x[0])),
+
+    def forward_gpu(self, x):
+        return cuda.cupy.sign(x[0]),
+
+    def backward(self, indexes, grad_outputs):
+        return grad_outputs[0] * 0,
```
Comment (okuta) [inline chainer/functions/math/sign.py]: Why don't you use `zeros_like`?

Comment (hvy) [reply chainer/functions/math/sign.py]: Thank you for the review. Don't we need to return a Variable object here?

Comment (okuta) [reply chainer/functions/math/sign.py]: Sorry, please return `None` in this case.

Comment (hvy) [reply chainer/functions/math/sign.py]: By doing so, double backprop won't work when this function is involved in the computational graph. Please correct me if I am wrong.

Comment (kmaehashi) [reply chainer/functions/math/sign.py]: I came across the same situation (just return zero) in #3539.  I'd like to know if there is a better way. (maybe ask @beam2d ?)

Comment (hvy) [reply chainer/functions/math/sign.py]: Both PyTorch and TensorFlow seem to return 0. This is the code for PyTorch. https://github.com/pytorch/pytorch/blob/3853d5da97ead031d7d77f39fe7ce2de243ce369/tools/autograd/derivatives.yaml#L436-L437

File: chainer/functions/__init__.py [hunk 2]
```diff
@@ -226,6 +226,8 @@
 from chainer.functions.math.prod import prod  # NOQA
 from chainer.functions.math.prod import Prod  # NOQA
 from chainer.functions.math.scale import scale  # NOQA
+from chainer.functions.math.sign import sign  # NOQA
+from chainer.functions.math.sign import Sign  # NOQA
```
Comment (niboshi) [inline chainer/functions/__init__.py]: Please remove `Sign`. We're not going to expose new function classes.

Comment (hvy) [reply chainer/functions/__init__.py]: Fixed.

**Comment by hvy** [comment]
After offline discussion, we've concluded that no class is needed for this function, and that the gradients should be `None` everywhere. I have changed the code accordingly.

**Comment by niboshi** [comment]
If we break the graph, `computational_graph` feature will not work.
Is it really acceptable?

**Comment by hvy** [comment]
Was this also considered? @okuta

**Comment by hvy** [comment]
After offline discussions, it was concluded that the current implementation behaves according to intended design.

**Comment by hvy** [comment]
Rebased on top of master.

File: tests/chainer_tests/functions_tests/math_tests/test_sign.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    @property
+    def no_grads(self):
+        return (False,)
+
+    def make_data(self):
+        raise NotImplementedError
+
+    def setUp(self):
+        self.eps = 1e-3
+        while True:
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_sign.py]: Is this necessary?

File: tests/chainer_tests/functions_tests/math_tests/test_sign.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+import chainer
+from chainer import cuda
+import chainer.functions as F
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+
+
+class UnaryFunctionsTestBase(unittest.TestCase):
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_sign.py]: How about merging UnaryFunctionsTestBase and TestSign

Comment (hvy) [reply tests/chainer_tests/functions_tests/math_tests/test_sign.py]: Merged the classes.

File: tests/chainer_tests/functions_tests/math_tests/test_sign.py (hunk 3 above)
Comment (hvy) [reply tests/chainer_tests/functions_tests/math_tests/test_sign.py]: No sorry, this is not necessary for the sign function. Removed!

**Comment by hvy** [comment]
@okuta san, thank you for the review. I fixed the code accordingly.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
LGTM!

**Comment by takagi** [comment]
I made a PR #3910 to adjust `convolution_nd`'s testing tolerance which makes CI fail.

**Comment by hvy** [comment]
@takagi san, thank you. I'll take a look at the tolerance fix..
